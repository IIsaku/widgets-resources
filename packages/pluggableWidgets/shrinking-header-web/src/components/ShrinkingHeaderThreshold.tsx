import { createElement, CSSProperties, ReactElement, ReactNode, useCallback, useEffect, useState } from "react";
import classNames from "classnames";
import { throttle } from "lodash-es";

import "../ui/ShrinkingHeader.scss";
import { useWrappingDivHeight } from "../utils/WrappingDivStyler";

export interface ShrinkingHeaderThresholdProps {
    rootElementRef?: (node: HTMLElement | null) => void;
    className?: string;
    style?: CSSProperties;
    tabIndex?: number;
    content?: ReactNode;
    shrinkThreshold: number;
}

export function ShrinkingHeaderThreshold(props: ShrinkingHeaderThresholdProps): ReactElement {
    const { rootElementRef, className, style, tabIndex, content, shrinkThreshold } = props;

    const [headerElement, setHeaderElement] = useState<HTMLElement>();
    const [shrunk, setShrunk] = useState(false);

    const actualClassName = classNames(
        "widget-shrinking-header",
        "widget-shrinking-header-threshold",
        {
            "widget-shrinking-header-shrunk": shrunk
        },
        className
    );

    const updateElement = useCallback(
        (node: HTMLElement | null) => {
            setHeaderElement(node ?? undefined);
            rootElementRef?.(node);
        },
        [rootElementRef, setHeaderElement]
    );

    useEffect(() => {
        const evaluateShrunkState = function (): void {
            if (window.scrollY >= shrinkThreshold) {
                setShrunk(true);
            } else {
                setShrunk(false);
            }
        };

        evaluateShrunkState();

        const onScroll = throttle(evaluateShrunkState, 250, { leading: true, trailing: true });

        document.addEventListener("scroll", onScroll);

        return () => {
            document.removeEventListener("scroll", onScroll);
        };
    }, [shrinkThreshold, setShrunk]);

    const wrappingDivHeight = useWrappingDivHeight(headerElement);

    return (
        <div className={actualClassName} style={{ ...style, height: wrappingDivHeight }} tabIndex={tabIndex}>
            <header ref={updateElement}>{content}</header>
        </div>
    );
}
