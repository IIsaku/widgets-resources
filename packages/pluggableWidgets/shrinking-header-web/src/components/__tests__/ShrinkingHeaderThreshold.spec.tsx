import { createElement } from "react";
import { act } from "react-dom/test-utils";
import { mount, ReactWrapper } from "enzyme";

import { ShrinkingHeaderThreshold, ShrinkingHeaderThresholdProps } from "../ShrinkingHeaderThreshold";
import { mockResizeObserver } from "../../utils/__tests__/ResizeObserverMock";

let lastMockedResizeObserver: ResizeObserver & { notifyChange: () => void };

mockResizeObserver(act, _lastMockedResizeObserver => {
    lastMockedResizeObserver = _lastMockedResizeObserver;
});

let documentEventListeners: Array<() => void>;

document.addEventListener = jest.fn().mockImplementation((_type: string, callback: () => void) => {
    documentEventListeners.push(callback);
});

describe("ShrinkingHeaderThreshold", () => {
    let defaultShrinkingHeaderProps: ShrinkingHeaderThresholdProps;

    beforeEach(() => {
        defaultShrinkingHeaderProps = {
            rootElementRef: jest.fn(),
            name: "name",
            className: "class-name",
            style: { color: "green" },
            tabIndex: 4,
            content: <h1>Header</h1>,
            shrinkThreshold: 250
        };

        documentEventListeners = [];
    });

    it("renders correctly when the page isn't scrolled past the threshold", () => {
        const shrinkingHeaderThreshold = mount(<ShrinkingHeaderThreshold {...defaultShrinkingHeaderProps} />);

        Object.defineProperty(shrinkingHeaderThreshold.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 100,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderThreshold.update();
        expect(shrinkingHeaderThreshold).toMatchSnapshot();
        expect(defaultShrinkingHeaderProps.rootElementRef).toHaveBeenCalledTimes(1);
    });

    function testShrinkClassBeingApplied(threshold: number): ReactWrapper {
        const shrinkingHeaderThreshold = mount(<ShrinkingHeaderThreshold {...defaultShrinkingHeaderProps} />);

        Object.defineProperty(shrinkingHeaderThreshold.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 100,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderThreshold.update();
        expect(documentEventListeners).toHaveLength(1);

        Object.defineProperty(window, "scrollY", { value: threshold, configurable: true });
        act(documentEventListeners[0]);
        Object.defineProperty(shrinkingHeaderThreshold.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 25,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderThreshold.update();
        expect(shrinkingHeaderThreshold).toMatchSnapshot();

        return shrinkingHeaderThreshold;
    }

    it("applies the shrink class when the page gets scrolled to the threshold", () => {
        testShrinkClassBeingApplied(250);
    });

    it("applies the shrink class when the page gets scrolled passed the threshold", () => {
        testShrinkClassBeingApplied(275);
    });

    it("removes the shrink class when the page gets scrolled back above the threshold", async () => {
        const shrinkingHeaderThreshold = testShrinkClassBeingApplied(275);

        Object.defineProperty(window, "scrollY", { value: 50, configurable: true });
        await new Promise(r => setTimeout(r, 250));
        act(documentEventListeners[0]);
        Object.defineProperty(shrinkingHeaderThreshold.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 100,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderThreshold.update();
        expect(shrinkingHeaderThreshold).toMatchSnapshot();
    });

    // == Unmount doesn't trigger the cleanup function of the useEffect. This could be caused by erroneous code in React 17 as with both Enzyme and React Testing Library the function isn't triggered, even when waiting for it.
    // it("unmounts correctly", () => {
    //     const spy = jest.spyOn(document, "removeEventListener");
    //
    //     const shrinkingHeaderLinear = mount(<ShrinkingHeaderLinear {...defaultShrinkingHeaderProps} />);
    //
    //     expect(documentEventListeners).toHaveLength(1);
    //
    //     shrinkingHeaderLinear.unmount();
    //     expect(spy).toHaveBeenCalledWith("scroll", documentEventListeners[0]);
    // });
});
