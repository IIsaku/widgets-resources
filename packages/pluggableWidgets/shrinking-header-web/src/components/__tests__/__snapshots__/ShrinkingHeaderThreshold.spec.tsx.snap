// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShrinkingHeaderThreshold applies the shrink class when the page gets scrolled passed the threshold 1`] = `
<ShrinkingHeaderThreshold
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  rootElementRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <header>
            <h1>
              Header
            </h1>
          </header>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  shrinkThreshold={250}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-threshold widget-shrinking-header-shrunk class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header>
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderThreshold>
`;

exports[`ShrinkingHeaderThreshold applies the shrink class when the page gets scrolled to the threshold 1`] = `
<ShrinkingHeaderThreshold
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  rootElementRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <header>
            <h1>
              Header
            </h1>
          </header>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  shrinkThreshold={250}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-threshold widget-shrinking-header-shrunk class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header>
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderThreshold>
`;

exports[`ShrinkingHeaderThreshold removes the shrink class when the page gets scrolled back above the threshold 1`] = `
<ShrinkingHeaderThreshold
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  rootElementRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <header>
            <h1>
              Header
            </h1>
          </header>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  shrinkThreshold={250}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-threshold widget-shrinking-header-shrunk class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header>
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderThreshold>
`;

exports[`ShrinkingHeaderThreshold removes the shrink class when the page gets scrolled back above the threshold 2`] = `
<ShrinkingHeaderThreshold
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  rootElementRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <header>
            <h1>
              Header
            </h1>
          </header>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  shrinkThreshold={250}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-threshold class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header>
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderThreshold>
`;

exports[`ShrinkingHeaderThreshold renders correctly when the page isn't scrolled past the threshold 1`] = `
<ShrinkingHeaderThreshold
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  rootElementRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <header>
            <h1>
              Header
            </h1>
          </header>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  shrinkThreshold={250}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-threshold class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header>
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderThreshold>
`;
